posthog/caching/warming.py

```diff
@@ -0,0 +1,139 @@
+from datetime import timedelta, UTC, datetime
+from collections.abc import Generator
+from typing import Optional
+
+import structlog
+from celery import shared_task
+from celery.canvas import chain
+from django.db.models import Q
+from prometheus_client import Counter
+from sentry_sdk import capture_exception
+
+from posthog.api.services.query import process_query_dict
+from posthog.caching.utils import largest_teams
+from posthog.clickhouse.query_tagging import tag_queries
+from posthog.errors import CHQueryErrorTooManySimultaneousQueries
+from posthog.hogql_queries.query_cache import QueryCacheManager
+from posthog.hogql_queries.legacy_compatibility.flagged_conversion_manager import conversion_to_query_based
+from posthog.hogql_queries.query_runner import ExecutionMode
+from posthog.models import Team, Insight, DashboardTile
+from posthog.schema import GenericCachedQueryResponse
+from posthog.tasks.utils import CeleryQueue
+
+logger = structlog.get_logger(__name__)
+
+STALE_INSIGHTS_COUNTER = Counter(
+    "posthog_cache_warming_stale_insights",
+    "Number of stale insights present",
+    ["team_id"],
+)
+PRIORITY_INSIGHTS_COUNTER = Counter(
+    "posthog_cache_warming_priority_insights",
+    "Number of priority insights warmed",
+    ["team_id", "dashboard", "is_cached"],
+)
+
+LAST_VIEWED_THRESHOLD = timedelta(days=7)
+
+
+def priority_insights(team: Team) -> Generator[tuple[int, Optional[int]], None, None]:
+    """
+    This is the place to decide which insights should be kept warm.
```

Comment:
Loose question: One of the goals of the `InsightCachingState` system was to make sure shared dashboards show fresh data – is this something already solved outside of cache warming, or might still be an issue?

Reply:
I'm still not sure... it depends on the use cases and goals for shared insights. It is solved partially by the fact that we now asynchronously refresh the insights if they are stale when viewed. So it's not needed anymore to prevent empty shared insights.
I can check if we also track when shared insights are viewed.. otherwise I wouldn't even want to refresh them all if we don't know that.